#!/usr/bin/env Rscript

# This script maps peak data as generated by, for example mm9RDtoGenes (so already mapped
# to the nearest genes) to differential expression data as generated by limma_xpn.R (so
# already annotated with genes)
# It basically just merges on Ensembl transcript name

library(IRanges)

options(stringsAsFactors = FALSE);
args <- commandArgs(trailingOnly=TRUE)

#these have already been mapped to their nearest transcript
peak.file <- args[1]
xpn.file <- args[2]
merge.file <- args[3]

peaks <- read.csv(peak.file)
xpn <- read.csv(xpn.file)


# the peaks are mapped to nearest ensembl gene IDs
# the expression data has Ensembl Gene ID annotation in the
# "Proportion_Ensembl_transcripts" field
ensembl.ids <- sub(".*(ENS.*)\\).*", "\\1", xpn[,"Proportion_Ensembl_transcripts"])
xpn<-cbind(xpn, ensembl.id=ensembl.ids)

# Chuck out anything that doesn't map to an Ensembl ID
peaks<-peaks[peaks[,"tss.feature"]!="",]
xpn<-xpn[xpn[,"ensembl.id"]!="",]

# Map between the 2 on the basis of Ensembl Gene ID.
merged <- merge(peaks, xpn, by.x="tss.feature", by.y="ensembl.id")

write.csv(merged, merge.file)




