#!/usr/bin/env Rscript

# This script takes a file as generated by, for example peaks2xpn.R, containing
# peak data mapped to the nearest TSS and generates some summmary plots and information
# Output goes to the directory that contains the input file

options(stringsAsFactors = FALSE)
args <- commandArgs(trailingOnly=TRUE)

filename <-  args[1]

#for testing
#filename <- "/mnt/data/NSC_chip_xpn_merge.csv"

outdir <- dirname(filename)





peak.data <- read.csv(filename) 

#  colnames(peak.data)
#  [1] "X"                              "tss.feature"                   
#  [3] "space"                          "start"                         
#  [5] "end"                            "width"                         
#  [7] "names"                          "values.Length"                 
#  [9] "values.Summit"                  "values.nTags"                  
# [11] "values.neg10log10pVal"          "values.FoldEnrichment"         
# [13] "values.FDR"                     "tss.strand"                    
# [15] "tss.start_position"             "tss.end_position"              
# [17] "tss.insideFeature"              "tss.distancetoFeature"         
# [19] "tss.shortestDistance"           "tss.fromOverlappingOrNearest"  
# [21] "mgi_symbol"                     "description"                   
# [23] "Chr"                            "Start"                         
# [25] "End"                            "Strand"                        
# [27] "IlluminaID"                     "logFC"                         
# [29] "AveExpr"                        "t"                             
# [31] "P.Value"                        "adj.P.Val"                     
# [33] "B"                              "Probe_id"                      
# [35] "Lumi_id"                        "Probe_sequence"                
# [37] "Probe_type"                     "Quality_score"                 
# [39] "RefSeq_transcripts"             "Proportion_RefSeq_transcripts" 
# [41] "UCSC_transcripts"               "Proportion_UCSC_transcripts"   
# [43] "GenBank_transcripts"            "Proportion_GenBank_transcripts"
# [45] "Exons"                          "Ensembl_transcripts"           
# [47] "Proportion_Ensembl_transcripts" "Lumi_transcriptomic_annotation"
# [49] "Lumi_transcriptomic_match"     

relevant <- c("names", "values.neg10log10pVal", "values.FoldEnrichment", "mgi_symbol", "IlluminaID", "logFC", "P.Value", "tss.distancetoFeature")

peak.mini <- peak.data[,relevant]


# Distribution of distances to peaks
# Really need to know what the random distribution is for this to be useful
# will have to annotate a set of randomly selected positions at some point.
outfile <- sub("\\.csv", "_distance_distribution.png", filename)
bmp(outfile)
plot(density(peak.mini[,"tss.distancetoFeature"]))
dev.off()


# distance to peak vs differential expression FC
outfile <- sub("\\.csv", "_distance_vs_FC.png", filename)
bmp(outfile)
plot(log10(abs(peak.mini[,"tss.distancetoFeature"])), peak.mini[,"P.Value"], pch=20)
dev.off()


# distance to peak vs differential expression p value
outfile <- sub("\\.csv", "_distance_vs_pval.png", filename)
bmp(outfile)
plot(log10(abs(peak.mini[,"tss.distancetoFeature"])), peak.mini[,"logFC"], pch=20)
dev.off()


# distance to peak vs logFC, with significant changes highlighted
outfile <- sub("\\.csv", "_distance_vs_FC_sig_highlight.png", filename)
bmp(outfile)
plot(peak.mini[,"tss.distancetoFeature"], peak.mini[,"logFC"], pch=20, xlim=c(-100000,100000))
sig <- which(peak.mini[,"P.Value"]<=0.05)
points(peak.mini[sig,"tss.distancetoFeature"], peak.mini[sig,"logFC"], pch=20, col="red", xlim=c(-100000,100000))
dev.off()             


# So, there's a relationship between the peaks and the differential expression of genes,
# as you might expect, but not a particularly strong one.

# There also seem to be a lot of peaks that are annotating as being closest to
# genes that are very far away. Are these just not real peaks? Look at the
# relationship between distance and peak enrichment?

outfile <- sub("\\.csv", "_distance_vs_peakFC.png", filename)
bmp(outfile)
plot(log10(abs(peak.mini[,"tss.distancetoFeature"])), peak.mini[,"values.FoldEnrichment"], pch=20)
dev.off()





  

# Maybe half of all peaks map to a know gene on the array. Which is probably enough to
# give us a global picture of what's going on?

#  esc <- read.csv("/mnt/data/ESC_chip_xpn_merge.csv")
#  nsc <- read.csv("/mnt/data/NSC_chip_xpn_merge.csv")
#  astro <- read.csv("/mnt/data/Astro_chip_xpn_merge.csv")
#  dim(esc)
# [1] 2693   49
#  dim(nsc)
# [1] 657  49
#  dim(astro)
# [1] 308  49


# esc.peaks <- read.csv("/mnt/esc/macs_300_1.0e-05/res_tss.csv")
# nsc.peaks <- read.csv("/mnt/nsc/macs_300_1.0e-05/res_tss.csv")
# astro.peaks <- read.csv("/mnt/astro/macs_300_1.0e-05/res_tss.csv")
# dim(esc.peaks)
# [1] 5172   21
# dim(nsc.peaks)
# [1] 1317   21
# dim(astro.peaks)
# [1] 677  21



